File: MinJu/231211_week/Baekjoon_1766_문제집.py
# 위상 정렬 
# 가능한 쉬운 문제부터 풀어야한다는 조건이 있으므로 우선순위 큐 
import heapq

n, m = map(int, input().split())
indegree = [0] * (n+1)
graph = [[] for _ in range(n+1)]
q = []

# 그래프 정보 + 진입 차수 저장
for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    indegree[b] += 1 

def topology_sort():

    # 진입 차수 0인 노드를 우선순위 큐에 삽입
    for i in range(1, n + 1):
        if indegree[i] == 0:
            heapq.heappush(q, i) 
    
    # 큐가 빌 때까지 반복
    while q:
        now = heapq.heappop(q)
        print(now, end=" ")

        # 해당 원소와 연결된 노드들의 진입차수에서 1 빼기
        for i in graph[now]:
            indegree[i] -= 1
            # 진입차수가 0이 되는 노드를 큐에 삽입
            if indegree[i] == 0:
                heapq.heappush(q, i)

topology_sort()

File: MyungJi/231218_week/baekjoon_1766_문제집.cpp
#include <bits/stdc++.h>
using namespace std;
int n, m, a, b;
vector<int> graph[100004];
int indegree[32004];
vector<int> result;
// 위상 정렬
void topologySort(){
    priority_queue<int, vector<int>, greater<int>> pq;
    for(int i=1;i<=n;i++){
        if (indegree[i]==0) pq.push(i);
    }
    while(!pq.empty()){
        int now = pq.top();
        pq.pop();
        result.push_back(now);
        for(int i=0;i<graph[now].size();i++){
            int node = graph[now][i];
            indegree[node]-=1;
            if(indegree[node]==0) pq.push(node);
        }

    }
}

int main(){
    // 1. 입력받기
    cin >> n >> m;
    for(int i=0;i<m;i++){
        cin >> a >> b;
        graph[a].push_back(b);
        indegree[b] += 1;
    }
    // 2. 위상정렬
    topologySort();
    // 3. 결과 출력
    for(int r:result) cout << r << " ";
}